### PR TITLE
Or 1566 thanos cannon add get message status for portals

### PR DESCRIPTION
**Pull request by nguyenzung**
Add getMessageStatus for Portals and examples 

Thank you!

### CODE REVIEW

File: packages/tokamak/sdk/src/portals.ts [hunk 1]
```diff
@@ -190,6 +196,65 @@ export class Portals {
...
+      )
+      const provenWithdrawal = await this.getProvenWithdrawal(
+        withdrawalMessageInfo.withdrawalHash
+      )
+      if (
+        provenWithdrawal === undefined ||
+        provenWithdrawal === null ||
+        provenWithdrawal.timestamp.toNumber() === 0
+      ) {
+        return MessageStatus.READY_TO_PROVE
```
Comment (0x6e616d) [inline packages/tokamak/sdk/src/portals.ts:234]:
```suggestion
      if (
        provenWithdrawal === undefined ||
        provenWithdrawal === null ||
        provenWithdrawal.timestamp.toNumber() === 0
      ) {
        return MessageStatus.READY_TO_PROVE
       }
       
       const BUFFER_TIME = 12
```

**Review by 0x6e616d — APPROVED** [review]
SGTM, but I want to make a little bit of a change. Can you refactor the `getL2ToL1MessageStatusByReceipt` function to prevent using the nested `if....else` by using the early return? I think It helps the code more clearly 

My suggestion:
```typescript
public async getL2ToL1MessageStatusByReceipt(
    txReceipt: TransactionReceipt
  ): Promise<MessageStatus> {
    const l2BlockNumber = txReceipt.blockNumber
    if (l2BlockNumber > (await this.getL2BlockNumberInOO())) {
      return MessageStatus.STATE_ROOT_NOT_PUBLISHED
    }
    const withdrawalMessageInfo = await this.calculateWithdrawalMessage(
      txReceipt
    )
    const provenWithdrawal = await this.getProvenWithdrawal(
      withdrawalMessageInfo.withdrawalHash
    )
    if (
      !provenWithdrawal ||
      provenWithdrawal.timestamp.toNumber() === 0
    ) {
      return MessageStatus.READY_TO_PROVE
    }

    const BUFFER_TIME = 12
    const currentTimestamp = Date.now()
    const finalizedPeriod = await this.getChallengePeriodSeconds()
    if (
      currentTimestamp / 1000 - BUFFER_TIME <
      provenWithdrawal.timestamp.toNumber() + finalizedPeriod
    ) {
      return MessageStatus.IN_CHALLENGE_PERIOD
    } 
    const finalizedStatus = await this.getFinalizedWithdrawalStatus(
      withdrawalMessageInfo.withdrawalHash
    )
    return finalizedStatus ? MessageStatus.RELAYED : MessageStatus.READY_FOR_RELAY
  }
```

**Comment by nguyenzung** [comment]
> ```ts
> public async getL2ToL1MessageStatusByReceipt(
>     txReceipt: TransactionReceipt
>   ): Promise<MessageStatus> {
>     const l2BlockNumber = txReceipt.blockNumber
>     if (l2BlockNumber > (await this.getL2BlockNumberInOO())) {
>       return MessageStatus.STATE_ROOT_NOT_PUBLISHED
>     }
>     const withdrawalMessageInfo = await this.calculateWithdrawalMessage(
>       txReceipt
>     )
>     const provenWithdrawal = await this.getProvenWithdrawal(
>       withdrawalMessageInfo.withdrawalHash
>     )
>     if (
>       !provenWithdrawal ||
>       provenWithdrawal.timestamp.toNumber() === 0
>     ) {
>       return MessageStatus.READY_TO_PROVE
>     }
> 
>     const BUFFER_TIME = 12
>     const currentTimestamp = Date.now()
>     const finalizedPeriod = await this.getChallengePeriodSeconds()
>     if (
>       currentTimestamp / 1000 - BUFFER_TIME <
>       provenWithdrawal.timestamp.toNumber() + finalizedPeriod
>     ) {
>       return MessageStatus.IN_CHALLENGE_PERIOD
>     } 
>     const finalizedStatus = await this.getFinalizedWithdrawalStatus(
>       withdrawalMessageInfo.withdrawalHash
>     )
>     return finalizedStatus ? MessageStatus.RELAYED : MessageStatus.READY_FOR_RELAY
>   }
> ```

Thank you, Nam
Let me use your code. It looks good!

**Comment by 0x6e616d** [comment]
Your work is amazing. Thank you 💯

**Review by rlgns98kr — APPROVED** [review]
Awesome!
I appreciate your fast and beautiful code!

After Nam's review, I think we can merge this.
Can you add the getMessageStatus function in here as well ? (🙇🏼‍♂️🥹)

**Review by 0x6e616d — APPROVED** [review]

**Comment by nguyenzung** [comment]
I missed using getMessageStatus for withdraw case. I updated!

**Pull request merged**